controllers/virtualmachinereplicaset/virtualmachinereplicaset_controller.go

```diff
@@ -202,38 +225,461 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Re
+					"FailedAdopt",
+					"Failed to adopt VirtualMachine %q: %v",
+					vm.Name,
+					err)
+
+				continue
+			}
+			ctx.Logger.Info("Adopted VirtualMachine", "vm", vm.Name)
+			r.Recorder.Eventf(ctx.ReplicaSet, corev1.EventTypeNormal, "SuccessfulAdopt", "Adopted VirtualMachine %q", vm.Name)
+		}
+
+		filteredMachines = append(filteredMachines, vm)
+	}
+
+	// If not already present, add a label specifying the VirtualMachineReplicaSet name to VirtualMachines.
+	for _, vm := range filteredMachines {
+		// Note: MustEqualValue is used here as the value of this label will be
+		// a hash if the MachineSet name is longer than 63 characters.
+		if msNameLabelValue, ok := vm.Labels[vmopv1.VirtualMachineReplicaSetNameLabel]; ok &&
+			MustEqualValue(ctx.ReplicaSet.Name, msNameLabelValue) {
+
+			continue
+		}
+
+		// Label not present, patch the VirtualMachine.
+		helper, err := patch.NewHelper(vm, r.Client)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to apply %s label to VirtualMachine %q",
+				vmopv1.VirtualMachineReplicaSetNameLabel, vm.Name)
+		}
+
+		// MustFormatValue is used here as the value of this label will be a
+		// hash if the VirtualMachineReplicaSet name is longer than max allowed
+		// label value length of 63 characters.
+		vm.Labels[vmopv1.VirtualMachineReplicaSetNameLabel] = MustFormatValue(ctx.ReplicaSet.Name)
+
+		if err := helper.Patch(ctx, vm); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to apply %s label to Machine %q",
+				vmopv1.VirtualMachineReplicaSetNameLabel, vm.Name)
+		}
+
+		// TODO: Propagate the Deployment label from Deployment to replica set to VirtualMachine if
+		// it is set on the replica set.
+	}
+
+	syncErr := r.syncReplicas(ctx, ctx.ReplicaSet, filteredMachines)
+
+	// Update the status of the VirtualMachineReplicaSet since syncing might
+	// have resulted in replicas being added or removed.
+	r.updateStatus(ctx, ctx.ReplicaSet, filteredMachines)
+
+	if syncErr != nil {
+		return ctrl.Result{}, errors.Wrapf(syncErr, "failed to sync MachineSet replicas")
+	}
+
+	var replicas int32
+	if ctx.ReplicaSet.Spec.Replicas != nil {
+		replicas = *ctx.ReplicaSet.Spec.Replicas
+	}
+
+	// Queue faster reconciles until all replicas are available.
+	if ctx.ReplicaSet.Status.ReadyReplicas != replicas {
+		return ctrl.Result{RequeueAfter: 15 * time.Second}, nil
+	}
+
+	return ctrl.Result{}, nil
 }
 
-func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineReplicaSetContext) (reterr error) {
-	if controllerutil.ContainsFinalizer(ctx.ReplicaSet, finalizerName) {
-		defer func() {
-			r.Recorder.EmitEvent(ctx.ReplicaSet, "Delete", reterr, false)
-		}()
+// MustFormatValue returns the passed inputLabelValue if it meets the standards for a Kubernetes label value.
+// If the name is not a valid label value this function returns a hash which meets the requirements.
+// Reuse the hashing used by CAPI for MachineSets.
+func MustFormatValue(str string) string {
```

Comment:
Given this function is a `Must` variant, I would rather see a `FormatValue` that returns `string, error` and then have `MustFormatValue` call that, and then using `panic` if `FormatValue` returns an error.

Reply:
So, looks like the writer exposed by the FNV hasher never returns an error.  https://cs.opensource.google/go/go/+/refs/tags/go1.22.3:src/hash/fnv/fnv.go;l=106

Still, I think the `Must` analogy and pattern applies here.  Moved to `util` package and followed your suggestion.  Thanks!